drivers: audio: Add get_caps API to DMIC, I2S, and Codec subsystems

michal-chvatal opened the pull request:
This PR introduces `get_caps()` API to DMIC, I2S, and Codec subsystems, enabling
drivers to report their audio capabilities such as supported sample rates, bit widths,
channel counts, buffer requirements, and frame timing.

It implements the API for MCUX I2S Flexcomm, DMIC, and WM8904 audio drivers.

It also introduces common audio definitions in audio_caps.h to unify capability
flags across subsystems.

Changes are related to new MediaPipe library.

This PR depends on: [I2s mcux flexcomm tx rx num buff to kconfig #108239](https://github.com/zephyrproject-rtos/zephyr/pull/108239)

michal-chvatal committed:
drivers: audio: Add get_caps API to DMIC, I2S, and Codec subsystems

Introduce `get_caps()` API to DMIC, I2S, and Codec subsystems to expose
driver-specific audio capabilities such as supported sample rates,
bit widths, channel counts, buffer requirements, and frame timing.

Introduce common audio definitions in `audio_caps.h` to unify capability
flags across subsystems.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>

michal-chvatal committed:
drivers: audio: Add get_caps API implementation

Implement the `get_caps()` API for MCUX I2S Flexcomm, DMIC, and
WM8904 audio drivers to expose driver-specific capabilities
including supported sample rates, bit widths, channel counts, and
buffer requirements.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>

dbaluta commented on drivers/audio/dmic_mcux.c:
```diff
@@ -18,6 +18,7 @@
 
```
Please do the code formatting in a separate patch. A patch should do one single logical change.

michal-chvatal committed:
drivers: audio: dmic_mcux: reflow lines to 100-column limit

Collapse multi-line function signatures and call sites in the MCUX
DMIC driver onto fewer lines within the 100-column limit, matching
clang-format output. No functional changes.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>

decsny commented on include/zephyr/drivers/i2s.h:
```diff
@@ -537,6 +540,25 @@ static inline int z_impl_i2s_trigger(const struct device *dev,
+	struct i2s_driver_api *api = (struct i2s_driver_api *)dev->api;
+
+	if (api->get_caps == NULL) {
+		return -ENOSYS;
```
this particular error code should be documented what it means

decsny reviewed (dismissed):
you need to add testing of this feature

michal-chvatal commented on drivers/audio/dmic_mcux.c:
```diff
@@ -651,76 +628,87 @@ static int dmic_mcux_read(const struct device *dev,
 	return 0;
 }
 
+static int dmic_mcux_get_caps(const struct device *dev, struct audio_caps *caps)
```
After a detailed private discussion with @ZhaoxiangJin, we concluded that while describing hardware capabilities via DT can be beneficial, especially when features differ across platforms, it’s not strictly necessary for all cases.

In particular, for this PR, we agreed to keep the current approach. This allows drivers to report static capabilities (like sample rates, bit widths, and channel counts) in a structured and generic way, without introducing additional complexity into the DT.

If future platforms introduce variations in hardware features (e.g., different number of PDM channels), those specific values can be added to DT and reused not only for get_caps() but also for other parts of the driver logic.

michal-chvatal committed:
tests: audio: Add get_caps API tests for I2S and DMIC drivers

Add test for the `get_caps()` API in both I2S and DMIC
driver test suites. Tests validate successful operation,
capability value ranges, and error handling with NULL parameters.

Tests handle drivers that don't implement get_caps by gracefully
skipping when -ENOSYS is returned, ensuring compatibility
across different driver implementations.

Signed-off-by: Michal Chvatal <michal.chvatal@nxp.com>

TomasBarakNXP commented:
@copilot resolve the merge conflicts in this pull request

zejiang0jason commented:
> @copilot resolve the merge conflicts in this pull request

does this work?

TomasBarakNXP commented:
> > @copilot resolve the merge conflicts in this pull request
> 
> does this work?

Doesn't seem so

TomasBarakNXP commented:
@anangl ping for review

hakehuang commented on include/zephyr/audio/codec.h:
```diff
@@ -674,6 +686,31 @@ static inline int audio_codec_register_done_callback(const struct device *dev,
 	return api->register_done_callback(dev, tx_cb, tx_cb_user_data, rx_cb, rx_cb_user_data);
 }
 
+/**
```
Copy-paste error: this says "Get i2s capabilities" but this is the codec API. Should be "Get audio codec capabilities".

hakehuang commented on drivers/audio/dmic_nxp_micfil.c:
```diff
@@ -614,10 +614,28 @@ static int nxp_micfil_init(const struct device *dev)
+
+	caps->min_total_channels = 1;
+	caps->max_total_channels = 8;
+	caps->supported_sample_rates = AUDIO_SAMPLE_RATE_16000;
```
`10000 µs = 10 ms`, not 1 ms. The comment says `/* 1ms minimum */` but the value is `10000`. Please fix either the value (to `1000`) or the comment (to `/* 10ms minimum */`). All other drivers in this PR use `1000` for 1 ms.

hakehuang commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -1066,6 +1066,28 @@ static int i2s_mcux_write(const struct device *dev, void *mem_block, size_t size
+static int i2s_mcux_get_caps(const struct device *dev, struct audio_caps *caps, enum i2s_dir dir)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(dir);
```
`min_total_channels = 1` contradicts the comment `/* Stereo minimum */`. If stereo is the minimum, use `2`. If mono is genuinely supported, remove the comment.

hakehuang commented on include/zephyr/audio/codec.h:
```diff
@@ -321,6 +323,12 @@ typedef int (*audio_codec_register_done_callback_t)(
 	const struct device *dev, audio_codec_tx_done_callback_t tx_cb, void *tx_cb_user_data,
 	audio_codec_rx_done_callback_t rx_cb, void *rx_cb_user_data);
 
+/**
```
Nit: the trailing comment `/* Include common audio caps */` is redundant — the filename is self-explanatory. Please remove it here and in `dmic.h` / `i2s.h`.

hakehuang commented on drivers/audio/dmic_mcux.c:
```diff
@@ -658,76 +634,85 @@ static int dmic_mcux_read(const struct device *dev,
 }
 
+static int dmic_mcux_get_caps(const struct device *dev, struct audio_caps *caps)
+{
```
These sample rates are hardcoded, but the achievable rates depend on the PDM bit clock which is read dynamically in `dmic_mcux_configure()`. Please add a comment clarifying these are static approximations, or derive them from the clock configuration.

hakehuang commented on drivers/i2s/i2s_mcux_sai.c:
```diff
@@ -1066,6 +1066,28 @@ static int i2s_mcux_write(const struct device *dev, void *mem_block, size_t size
 	return ret;
 }
 
```
`dir` is marked `ARG_UNUSED` here, but in `i2s_mcux_flexcomm.c` it is used to select the correct `min_num_buffers`. Please either use `dir` for SAI as well, or add a comment explaining why TX and RX caps are identical for this controller.

hakehuang reviewed (changes requested):
Thank you for this patch — adding a `get_caps()` API across DMIC, I2S, and Codec subsystems is a useful capability. I have several issues that need to be addressed before this can be merged.

## Blocking issues

### 1. Wrong docstring in `audio_codec_get_caps()` (`include/zephyr/audio/codec.h`)
The inline doc says **"Get i2s capabilities"** — this is a copy-paste error. It should say **"Get audio codec capabilities"**.

### 2. `dmic_nxp_micfil.c`: `min_frame_interval` value/comment mismatch
```c
caps->min_frame_interval = 10000;  /* 1ms minimum */
```
`10000 µs = 10 ms`, not 1 ms. Every other driver in this PR uses `1000` for 1 ms. Either fix the value to `1000` or fix the comment to `/* 10ms minimum */`.

### 3. `i2s_mcux_sai.c`: contradictory `min_total_channels` and comment
```c
caps->min_total_channels = 1; /* Stereo minimum */
```
`1` and "Stereo minimum" are contradictory. If stereo is the minimum, use `2`. If mono is genuinely supported, remove the comment.

### 4. Unresolved dependency on #108239
This PR depends on #108239 which is not yet merged. Please do not merge until that dependency lands.

## Should-fix issues

### 5. `dmic_mcux.c`: hardcoded sample rates may not reflect runtime capability
`supported_sample_rates` is set to `AUDIO_SAMPLE_RATE_16000 | AUDIO_SAMPLE_RATE_48000` statically, but the actual achievable rates depend on the PDM bit clock read dynamically in `dmic_mcux_configure()`. Please add a code comment making clear these are static approximations, or derive them from the clock rate.

### 6. `i2s_mcux_sai.c`: `dir` parameter silently ignored
Both `dev` and `dir` are marked `ARG_UNUSED`, meaning TX and RX always return identical caps. In `i2s_mcux_flexcomm.c` the `dir` parameter is actually used to select the correct buffer count. Please either use `dir` in SAI as well, or add a comment explaining why TX/RX caps are identical for this controller.

### 7. Trailing comments on `#include` lines
```c
#include <zephyr/audio/audio_caps.h> /* Include common audio caps */
```
The comment adds no information — the filename is self-explanatory. Please remove it from `codec.h`, `dmic.h`, and `i2s.h`.

### 8. `wm8962.c`: `DEVICE_API` initializer formatting
The reformatted `DEVICE_API` struct initializer keeps the opening brace on the same line as `.configure`, which differs from the style used in `wm8904.c` and `da7212.c`. Please align with the rest of the file.

## Suggestions (optional but recommended)

### 9. Consider adding `max_num_buffers` to `struct audio_caps`
Currently only `min_num_buffers` is present. For drivers where TX and RX buffer counts differ, callers have no way to know the upper bound. Consider adding `max_num_buffers` or documenting that `min == max` is the intended contract.

### 10. `struct audio_caps` missing Doxygen group membership
The `@defgroup` blocks for bit widths and sample rates are defined, but `struct audio_caps` itself is not placed in any group. This makes the Doxygen output inconsistent.

### 11. `supported_sample_rates` bit headroom
With `BIT(25)` already used (`AUDIO_SAMPLE_RATE_768000`), only 6 bits remain in `uint32_t`. Consider `uint64_t` or document the intentional limit.

---

Please also confirm that the formatting-only changes (flagged by @dbaluta) are in a separate commit from the functional changes.

hakehuang commented on drivers/audio/dmic_mcux.c:
```diff
@@ -658,76 +634,90 @@ static int dmic_mcux_read(const struct device *dev,
 }
 
+static int dmic_mcux_get_caps(const struct device *dev, struct audio_caps *caps)
+{
```
Still missing the clarifying comment requested in the previous review. Please add a note that these sample rates are static approximations and the actual supported rates depend on the PDM bit clock configured at runtime in `dmic_mcux_configure()`.

TomasBarakNXP replied:
Exactly this comment is on lines 644-646.

hakehuang commented on include/zephyr/audio/dmic.h:
```diff
@@ -383,6 +395,31 @@ static inline int dmic_read(const struct device *dev, uint8_t stream,
+	const struct dmic_driver_api *api = DEVICE_API_GET(dmic, dev);
+
+	if (api->get_caps == NULL) {
+		return -ENOSYS;
```
(@decsny's comment) Please resolve @decsny's inline comment — the `-ENOSYS` retval description is present in the docstring but the original inline comment thread was dismissed without acknowledgement. Please mark it resolved or reply.

TomasBarakNXP replied:
The comment above is resolved.

hakehuang reviewed (changes requested):
Thank you for addressing the previous review comments. Most blocking issues are resolved. Two items remain.

## Still needs fixing

### 1. `dmic_mcux.c`: hardcoded sample rates — no clarifying comment added
The previous review asked for a comment clarifying that `AUDIO_SAMPLE_RATE_16000 | AUDIO_SAMPLE_RATE_48000` are static approximations, since the actual achievable rates depend on the PDM bit clock read dynamically in `dmic_mcux_configure()`. This was not addressed. Please add a comment such as:
```c
/*
 * These are static approximations. Actual supported rates depend on
 * the PDM bit clock configured at runtime in dmic_mcux_configure().
 */
```

### 2. Unresolved dependency on #108239
This PR still depends on #108239 which is not yet merged. Do not merge until that lands.

## Previously raised by @decsny — still unresolved

The `-ENOSYS` return value in `dmic_get_caps()` and `i2s_get_caps()` is not documented with a description of what it means. Please update the `@retval` line to:
```c
 * @retval -ENOSYS if the driver does not implement get_caps
```
(already present — but `@decsny`'s inline comment on the original lines was dismissed without a reply. Please mark it resolved.)

## Resolved — confirmed fixed ✓

- `audio_codec_get_caps()` docstring corrected to "Get audio codec capabilities"
- `dmic_nxp_micfil.c` `min_frame_interval` fixed to `1000` (1 ms)
- `i2s_mcux_sai.c` `min_total_channels` comment corrected to "Mono minimum"
- Redundant `/* Include common audio caps */` inline comments removed
- `wm8962.c` `DEVICE_API` initializer reformatted consistently
- `i2s_mcux_sai.c` `ARG_UNUSED(dir)` now has an explanatory comment

hakehuang commented:
Dependency #108239 (_I2s mcux flexcomm tx rx num buff to kconfig_) was merged on 2026-06-29. That blocker is now cleared. Remaining items before this PR can merge:

1. Add a comment in `dmic_mcux.c` clarifying that the hardcoded sample rates are static approximations (dependent on PDM bit clock at runtime).
2. Acknowledge/resolve @decsny's `-ENOSYS` documentation comment.

hakehuang reviewed (approved):
board test pass on v4.4.0-6930-gc3c0b559f1a1

TomasBarakNXP commented:
> Dependency #108239 (_I2s mcux flexcomm tx rx num buff to kconfig_) was merged on 2026-06-29. That blocker is now cleared. Remaining items before this PR can merge:
> 
> 1. Add a comment in `dmic_mcux.c` clarifying that the hardcoded sample rates are static approximations (dependent on PDM bit clock at runtime).
> 2. Acknowledge/resolve @decsny's `-ENOSYS` documentation comment.

Both items are addressed.

TomasBarakNXP commented:
@anangl ping for review, thanks.
